### PR TITLE
[CVCUDA] Utilize CV-CUDA batch processing function

### PR DESCRIPTION
**Pull request by wang-xinyu**
### PR types(PR类型)

CV-CUDA

### Description

- NormalizeAndPermute CUDA kernel support batch processing
- Store cache in Manager instead of Processor
- Implement basics of CV-CUDA batch processing
- ResizeByShort, CenterCrop support CV-CUDA batch processing
- PaddleClas preprocessor support CV-CUDA batch processing

### CODE REVIEW

**Comment by wang-xinyu** [comment]
Tested accuracy on the CI data(50 images) with CI script https://github.com/felixhjh/fastdeploy_ci/blob/main/scripts_ci/test_resnet50_vd.py.

When using CV-CUDA, the top1 and top5 accuracy are same as OpenCV.

But the scores are slightly different.

OpenCV | CV-CUDA
image

**Comment by wang-xinyu** [comment]
Tested the model Predict() function, warmup 100 rounds，tested 1000 rounds and get avg. latency.
CPU: Intel(R) Xeon(R) Gold 6271C CPU @ 2.60GHz
GPU: T4
Model: ResNet50 224x224
Backend: TensorRT FP16

| Image Shape | Batch Size | OpenCV | CV-CUDA | Gain | CV-CUDA(Resize on CPU) | Gain |
|-|-|-|-|-|-|-|
| 640x480 | 1 | 1.81578ms | 1.46613ms |  **19.3%** | 1.54365ms | 15.0% |
| 640x480 | 4 | 5.1151ms | 3.47895ms | **32.0%** | 4.11307ms | 19.6% |
| 640x480 | 16 | 19.7402ms | 12.4244ms | **37.1%** | 14.5227ms | 26.4% |
| 1920x1080 | 1 | 2.08784ms | 2.18011ms | -4.4% | 1.73418ms | **16.9%** |
| 1920x1080 | 4 | 6.46092ms | 6.85611ms | -6.1% | 5.05582ms | **21.8%** |
| 1920x1080 | 16 | 23.3599ms | 26.7941ms | -14.7% | 17.6061ms | **24.6%** |

When the first operator is Resize, and input image size is large, maybe it's better to run resize on CPU, because the HostToDevice memcpy is time consuming.

Call below API to run initial resize on CPU.
```
Python:
model.preprocessor.use_cuda(enable_cv_cuda=True, gpu_id=0)
model.preprocessor.initial_resize_on_cpu(True)

C++:
model.GetPreprocessor().UseCuda(true, 0);
model.GetPreprocessor().InitialResizeOnCpu(true);
```

File: fastdeploy/vision/classification/ppcls/ppcls_pybind.cc [hunk 1]
```diff
@@ -40,8 +40,10 @@ void BindPaddleClas(pybind11::module& m) {
            })
       .def("use_cuda",
            [](vision::classification::PaddleClasPreprocessor& self,
```
Comment (jiangjiajun) [inline fastdeploy/vision/classification/ppcls/ppcls_pybind.cc]: 这个是继承的基类的函数，应该给基类绑定一个python接口，这样之后改造的各模型就无需再单独绑定`use_cuda`这个函数了

File: fastdeploy/vision/classification/ppcls/preprocessor.cc [hunk 2]
```diff
@@ -100,32 +100,23 @@ void PaddleClasPreprocessor::DisablePermute() {
...
-        FDERROR << "Failed to processs image:" << i << " in "
-                << processors_[j]->Name() << "." << std::endl;
-        return false;
-      }
+  for (size_t j = 0; j < processors_.size(); ++j) {
+    ProcLib lib = ProcLib::DEFAULT;
+    if (initial_resize_on_cpu_ && j == 0 &&
+        processors_[j]->Name().find("Resize") == 0) {
+      lib = ProcLib::OPENCV;
+    }
```
Comment (jiangjiajun) [inline fastdeploy/vision/classification/ppcls/preprocessor.cc:110]: 这个initial_resize_on_cpu_的参数是属于基类的，是否能够放到基类中去控制。否则后面所有继承这个基类的预处理，都得自行处理intialize_resize_on_cpu_的逻辑，影响预处理模块的通用性

File: fastdeploy/vision/common/processors/base.h [hunk 3]
```diff
@@ -51,41 +52,58 @@ class FASTDEPLOY_DECL Processor {
...
+    return true;
+  }
+
   virtual bool ImplByCvCuda(Mat* mat) {
     return ImplByOpenCV(mat);
   }
 
+  virtual bool ImplByCvCuda(MatBatch* mat_batch) {
+    for (size_t i = 0; i < mat_batch->mats->size(); ++i) {
+      if (ImplByCvCuda(&(*(mat_batch->mats))[i]) != true) {
```
Comment (jiangjiajun) [inline fastdeploy/vision/common/processors/base.h:96]: 这里看起来仍然是按照单张来处理的吗？

**Review by jiangjiajun — CHANGES_REQUESTED** [review]

File: fastdeploy/vision/common/processors/base.h (hunk 3 above)
Comment (wang-xinyu) [reply fastdeploy/vision/common/processors/base.h:96]: 这里是基类的函数，如果没有重载，就按单张来处理的

File: fastdeploy/vision/classification/ppcls/ppcls_pybind.cc (hunk 1 above)
Comment (wang-xinyu) [reply fastdeploy/vision/classification/ppcls/ppcls_pybind.cc]: 有道理，我改一下

File: fastdeploy/vision/common/processors/resize_by_short.cc [hunk 4]
```diff
@@ -89,32 +89,77 @@ bool ResizeByShort::ImplByFlyCV(Mat* mat) {
...
+  nvcv::ImageBatchVarShape dst_batch(mat_batch->mats->size());
+  CreateCvCudaImageBatchVarShape(dst_tensors, dst_batch);
+
+  // CV-CUDA Interp value is compatible with OpenCV
+  cvcuda::Resize resize_op;
+  resize_op(mat_batch->Stream(), src_batch, dst_batch,
+            NVCVInterpolationType(interp_));
+
+  for (size_t i = 0; i < mat_batch->mats->size(); ++i) {
+    Mat* mat = &(*(mat_batch->mats))[i];
```
Comment (jiangjiajun) [inline fastdeploy/vision/common/processors/resize_by_short.cc]: Mat改名为FDMat, MatBatch改名为FDMatBatch

**Review by jiangjiajun — APPROVED** [review]

**Pull request merged**